src/post_processing.jl

```diff
@@ -730,9 +730,83 @@ function update_system!(sys::PSY.System, data::PowerFlowData; time_step = 1)
+
+Compute the penalty factors (active power loss factors) and store the result in `destination`, which is the view in the `loss_factors`` matrix of the `PowerFlowData``.
+The loss factors are computed using the Jacobian matrix `J` and the vector of partial derivatives of 
+slack power with respect to bus voltages (angle and magnitude) `dSbus_dV_ref`.
+The approach is interpreting the change in slack active power injection as the change of the grid active power losses.
+The function uses the KLU library for sparse matrix factorization to calculate the loss factors.
+The loss factors are a linear approximation of the change in slack active power injection respect to the change in active power injections at each bus.
+
+# Arguments
+- `J::SparseMatrixCSC{Float64, Int64}`: The Jacobian matrix in sparse format.
+- `dSbus_dV_ref::Vector{Float64}`: The reference vector for the change in slack bus power with respect to bus voltages (PV, PQ buses for voltage angle, PQ buses for voltage magnitude).
+- `destination::SubArray{Float64}`: The view in the penalty factors matrix where the computed penalty factors will be stored.
+
+"""
+function penalty_factors!(
+    J::SparseMatrixCSC{Float64, Int64},
+    dSbus_dV_ref::Vector{Float64},
+    destination::SubArray{Float64},
+)
+    J_t = SparseMatrixCSC(transpose(J))
+    # fact = LinearAlgebra.factorize(J_t)
+    # note: KLU only operates on square matrices
+    fact = KLU.klu(J_t)
+    f = fact \ dSbus_dV_ref
+    destination .= f[1:length(destination)]  # we are only interested in the active power loss factors
+end
+
+"""
+    penalty_factors_brute_force(data::PowerFlowData; kwargs...)
+
+Calculate the penalty factors for each bus in the power flow data using a brute force method.
+This function calculates the penalty factors for each bus in the power flow data by perturbing the active power injection 
+at each bus by a small step size and solving the power flow equations. 
+The loss factor value is computed as the change in the reference bus power injection divided by the step size.
+
+# Arguments
+- `data::PowerFlowData`: The power flow data containing bus types, active power injections, and other relevant information.
+- `step_size::Float64 = 1e-6`: The step size used to perturb the active power injection at each bus.
+- `kwargs...`: Additional keyword arguments to be passed to the `solve_powerflow!` function.
+
+# Returns
+- `loss_factors::Array{Float64, 2}`: A 2D array of penalty factors for each bus and time step.
+
+# Notes
+- The reference bus type is assumed to remain constant between time steps.
+- The initial power flow solution is computed to establish the starting power injection at the slack bus.
+"""
+
+function penalty_factors_brute_force(
+    data::PowerFlowData;
+    step_size::Float64 = 1e-6,
+    kwargs...,
+)
+
+    # we assume that the bus type for ref bus does not change between time steps
+    ref, _, _ = bus_type_idx(data, 1)
```

Comment:
Aren't we doing a lot of extra work here that we're throwing away?

Reply:
Could have `bus_type_idx` take a kwarg for what bus types to calculate that defaults to all three of them

Reply:
good point

Reply:
(or a positional argument if that's more performant than a keyword argument)